auth/constructor.go

```diff
@@ -0,0 +1,191 @@
+package auth
+
+import (
+	"fmt"
+
+	"github.com/xmidt-org/themis/xlog"
+	"github.com/xmidt-org/webpa-common/basculechecks"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	gokitprometheus "github.com/go-kit/kit/metrics/prometheus"
+	"github.com/justinas/alice"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/xmidt-org/bascule"
+	"github.com/xmidt-org/bascule/basculehttp"
+	"github.com/xmidt-org/bascule/key"
+	"github.com/xmidt-org/webpa-common/basculemetrics"
+	"go.uber.org/fx"
+)
+
+type PrimaryCOptionsIn struct {
+	fx.In
+	Logger  log.Logger
+	Options []basculehttp.COption `group:"primary_bascule_constructor_options"`
+}
+
+type PrimaryTokenFactoryIn struct {
+	fx.In
+	Logger       log.Logger
+	DefaultKeyID string         `name:"primary_bearer_default_kid"`
+	Resolver     key.Resolver   `name:"primary_bearer_key_resolver" optional:"true"`
+	Leeway       bascule.Leeway `name:"primary_bearer_leeway"`
+}
+
+type PrimaryBasculeMetricListenerIn struct {
+	fx.In
+	Listener *basculemetrics.MetricListener `name:"primary_bascule_metric_listener"`
+}
+
+func ProvidePrimaryBasculeConstructor() fx.Option {
+	return fx.Provide(
+		fx.Annotated{
+			Group: "primary_bascule_constructor_options",
+			Target: func(in PrimaryBasculeMetricListenerIn) basculehttp.COption {
+				return basculehttp.WithCErrorResponseFunc(in.Listener.OnErrorResponse)
+			},
+		},
+		fx.Annotated{
+			Group: "primary_bascule_constructor_options",
+			Target: func(in PrimaryTokenFactoryIn) basculehttp.COption {
+				in.Logger.Log(level.Key(), level.DebugValue(), xlog.MessageKey(), "providing token factory option", "server", "primary")
+				if in.Resolver == nil {
+					in.Logger.Log(level.Key(), level.WarnValue(), xlog.MessageKey(), "providing nil token factory option as resolver was not defined", "server", "primary")
+					return nil
+				}
+				return basculehttp.WithTokenFactory("Bearer", basculehttp.BearerTokenFactory{
+					DefaultKeyId: in.DefaultKeyID,
+					Resolver:     in.Resolver,
+					Parser:       bascule.DefaultJWTParser,
+					Leeway:       in.Leeway,
+				})
+			},
+		},
+		fx.Annotated{
+			Group: "primary_bascule_constructor_options",
+			Target: func(in primaryProfileIn) (basculehttp.COption, error) {
+				if in.Profile == nil {
+					return nil, nil
+				}
+				basicTokenFactory, factoryErr := basculehttp.NewBasicTokenFactoryFromList(in.Profile.Basic)
```

Comment:
in the old auth setup, we checked if there were any valid basic auths before adding the token factory.  This bascule function doesn't check that, so doesn't return an error in that case.  We should either add this in bascule or have a check here and not add the token factory if there are no basic auths.

This doesn't affect the api result (we return 403 either way), but does affect the log statement and metrics ("key not supported" vs "failed to parse and validate token" with "invalid password").

Reply:
Good point! I'll add a check 👍